apollo-federation/src/sources/connect/json_selection/README.md

```diff
@@ -574,25 +637,77 @@ registered in the `JSONSelection` parser in order to be recognized.
 For the time being, only a fixed set of known methods are supported, though this
 list may grow and/or become user-configurable in the future:
 
-> Full disclosure: even this list is still aspirational, but suggestive of the
-> kinds of methods that are likely to be supported in the next version of the
-> `JSONSelection` parser.
-
 ```graphql
-list->first { id name }
-list->last.name
-list->slice($args.start, $args.end)
-list->reverse
-some.value->times(2)
-some.value->plus($addend)
-some.value->minus(100)
-some.value->div($divisor)
-isDog: kind->eq("dog")
-isNotCat: kind->neq("cat")
-__typename: kind->match({ "dog": "Dog", "cat": "Cat" })
-decoded: utf8Bytes->decode("utf-8")
-utf8Bytes: string->encode("utf-8")
-encoded: bytes->encode("base64")
+# The ->echo method returns its first input argument as-is, ignoring
+# the input data. Useful for embedding literal values, as in
+# $->echo("give me this string"), or wrapping the input value.
+__typename: $->echo("Book")
+wrapped: field->echo({ fieldValue: @ })
+
+# Returns the type of the data as a string, e.g. "object", "array",
+# "string", "number", "boolean", or "null". Note that `typeof null` is
+# "object" in JavaScript but "null" for our purposes.
+typeOfValue: value->typeof
+
+# When invoked against an array, ->map evaluates its first argument
+# against each element of the array, binding the element values to `@`,
+# and returns an array of the results. When invoked against a non-array,
+# ->map evaluates its first argument against that value and returns the
+# result without wrapping it in an array.
+doubled: numbers->map(@->mul(2))
+types: values->map(@->typeof)
+
+# Returns true if the data is deeply equal to the first argument, false
+# otherwise. Equality is solely value-based (all JSON), no references.
+isObject: value->typeof->eq("object")
+
+# Takes any number of pairs [candidate, value], and returns value for
+# the first candidate that equals the input data. If none of the
+# pairs match, a runtime error is reported, but a single-element
+# [<default>] array as the final argument guarantees a default value.
+__typename: kind->match(
+    ["dog", "Canine"],
+    ["cat", "Feline"],
+    ["Exotic"]
+)
+
+# Like ->match, but expects the first element of each pair to evaluate
+# to a boolean, returning the second element of the first pair whose
+# first element is true. This makes providing a final catch-all case
+# easy, since the last pair can be [true, <default>].
+__typename: kind->matchIf(
+    [@->eq("dog"), "Canine"],
+    [@->eq("cat"), "Feline"],
+    [true, "Exotic"]
```

Comment:
`matchIf` actually adds a little extra complexity to language evaluation in that the type of the individual elements of arrays matter. So now the validation has to check that each array is `[bool | () -> bool, Value]`. If we want to make function handling general-purpose (the main idea, I think behind putting control flow in functions), type signatures are gonna get pretty hairy.

Reply:
The return type just becomes a union over the possible `Value` types, no?

Or we could just say (for the time being) that the method's return type will always be a JSON value, and iteratively improve the type system story over time. The only thing you might lose is the safety of sub-selecting the result of the `->matchIf` function, but that's often fine because the result is a leaf value or scalar anyway.

Reply:
But it would be invalid to put a function on the left side which returns anything other than `bool`, right? So either we need to start storing type signatures _now_, or have validation/LSP have an allow-list of _every_ method which can be in that spot to catch errors.

I suppose the other option is to add truthiness to every value, but that's easy to get wrong as a user.

Reply:
I'm not sure if we should really allow for match to return varying types. Typescript gets around this by having type guards and other ways to narrow down unions that might be hard to port over to JSONSelection.

Also, I'd advise against implicit truthiness as it spawns a whole set of issues e.g. in JavaScript `{}` is `true` but `{asValue: false}` is not.

Reply:
I think we want a `->match` method capable of returning any JSON value, as a runtime capability, but of course we can impose additional type expectations when we have an enforceable type system.

My sense is that our type system will need to represent unions, and if `->match` can return multiple JSON values, then its return type should be the union of those possible return types. You're right that we may not be able to narrow down the union as effectively as TypeScript could, but if the developer chooses to return a single type across all branches, then the union type would devolve/simplify into just that type, as desired.

Even a complicated union of types might have a common field like `kind` shared across all the types, so it's not useless to be able to represent that situation accurately.